### PR TITLE
osd: Fix a bunch of stretch peering issues

### PR DESCRIPTION
**Pull request by gregsfortytwo**
A bug report came in on stretch mode; fixing and testing that
resulted in a number of small but very important bug fixes becoming evident.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/osd_types.h [hunk 1]
```diff
@@ -1517,10 +1517,11 @@ struct pg_pool_t {
     return peering_crush_bucket_count != 0;
   }
 
-  bool stretch_set_can_peer(const set<int>& want, const OSDMap& osdmap,
+  bool stretch_set_can_peer(const set<int>& want, const OSDMap *osdmap,
```
Comment (athanatos) [inline src/osd/osd_types.h:1520]: Why this change?  It appears that osdmap is assumed to be non-null, so normally a ref would be the order of the day?

File: src/osd/PeeringState.cc [hunk 2]
```diff
@@ -2406,13 +2413,15 @@ bool PeeringState::choose_acting(pg_shard_t &auth_log_shard_id,
...
   }
-  while (want.size() > pool.info.size) {
-    // async recovery should have taken out as many osds as it can.
-    // if not, then always evict the last peer
-    // (will get synchronously recovered later)
-    psdout(10) << __func__ << " evicting osd." << want.back()
-               << " from oversized want " << want << dendl;
-    want.pop_back();
+  if (!pool.info.is_stretch_pool()) {
+    while (want.size() > pool.info.size) {
```
Comment (athanatos) [inline src/osd/PeeringState.cc:2417]: I don't really understand this.  This is here because we choose to leave the extra want set osds in until async recovery makes its selections (22c8cda).  Is there something later on that handles this if we calc_replicated_acting_stretch includes extra osds?  Can calc_replicated_acting_stretch include extra osds?

File: src/osd/PeeringState.cc [hunk 3]
```diff
@@ -2651,7 +2650,8 @@ void PeeringState::activate(
 
   if (is_primary()) {
     // only update primary last_epoch_started if we will go active
-    if (acting.size() >= pool.info.min_size) {
+    if ((acting.size() >= pool.info.min_size) &&
+	 pool.info.stretch_set_can_peer(acting, get_osdmap().get(), NULL)) {
```
Comment (athanatos) [inline src/osd/PeeringState.cc:2654]: The repeated acting.size() >= pool.info.min_size was already not great.  With the added stretch_set_can_peer condition, I'd like some kind of helper, maybe acting_set_writeable?

File: src/osd/osd_types.h (hunk 1 above)
Comment (gregsfortytwo) [reply src/osd/osd_types.h:1520]: There was something in a dropped commit that made it easier this way, but that's not true any more and you're right. Changed!

File: src/osd/PeeringState.cc (hunk 2 above)
Comment (gregsfortytwo) [reply src/osd/PeeringState.cc:2417]: Ah, you are right! I misunderstood what was going on here when I made that change and should have looked more closely. I removed it and replaced it with a more verbose comment on the logic.

File: src/osd/PeeringState.cc (hunk 3 above)
Comment (gregsfortytwo) [reply src/osd/PeeringState.cc:2654]: I thought about this but it's still more lines of code to do a helper function than to condense down the three places we check.

Comment (athanatos) [reply src/osd/PeeringState.cc:2654]: Lines of code is not the concern.  The concern is that we'll fail to update one of these if we have to modify a condition.  I really do want a helper here.

**Comment by athanatos** [comment]
@gregsfortytwo I think you still need to push the new branch?

File: src/osd/PeeringState.cc (hunk 3 above)
Comment (gregsfortytwo) [reply src/osd/PeeringState.cc]: Fair enough.

**Comment by gregsfortytwo** [comment]
@athanatos pushed now; in addition to your comments note the extra patch swapping the boolean direction on adding a mandatory member.

**Review by athanatos — APPROVED** [review]

**Comment by gregsfortytwo** [comment]
https://pulpito.ceph.com/gregf-2021-03-13_08:24:54-rados-wip-stretch-fixes-312-distro-basic-gibba/

I didn't filter anything out, and between some messy jobs and a short-lived bug in master there were 34 failures and 4 dead jobs against 484 passes, but they all have a pretty clear fault that isn't related to the changes here.

5961449: Certificate verification failed (Suse certificate verification failure)
5961455: The script only supports CentOS
5961488: iscsi not valid:  iscsi not one of 'true', 'false'
5961497: Certificate verification failed
5961504: Certificate verification failed
5961511: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961512: This is getting an Aborted signal in the bstore_kv_sync thread
5961548: iscsi not valid:  iscsi not one of 'true', 'false'
5961564: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961585: The script only supports CentOS
5961587: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961602: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961613: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961615: iscsi not valid:  iscsi not one of 'true', 'false'
5961628: This fails on “Scrubbing terminated -- not all pgs were active and clean.”, but the ceph.log shows it was in “1 active+remapped+backfill_wait+backfill_toofull”
5961649: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961663: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961672: iscsi not valid:  iscsi not one of 'true', 'false'
5961681: Failed test_cephfs_mirror
5961718: The script only supports CentOS
5961723: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961756: iscsi not valid:  iscsi not one of 'true', 'false'
5961777: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961806: iscsi not valid:  iscsi not one of 'true', 'false'
5961833: The script only supports CentOS
5961873: iscsi not valid:  iscsi not one of 'true', 'false'
5961881: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961887: mount.nfs: Protocol not supported; ERROR: test_create_and_delete_export (tasks.cephfs.test_nfs.TestNFS)
5961891: This fails on “Scrubbing terminated -- not all pgs were active and clean.” at 00:40:43.341, but looking at ceph.log it shows PGs went from 1 remapped to all active+clean at 00:40:41.668994
5961933: https://tracker.ceph.com/issues/49726 FAILED ceph_assert(!version || comp->get_version64() == version)
5961940: iscsi not valid:  iscsi not one of 'true', 'false'
5961945: not sure exactly what this is, but it’s something to do with setup in cephadm
5961949: AssertionError: machine gibba037.front.sepia.ceph.com is locked by scheduled_teuthology@teuthology, not scheduled_gregf@teuthology
Dead jobs:
5961645: Maybe osd.5 crashed, but I can’t tell because we don’t have logs? :( Something went wrong for the supervisor as well so I think it was just bad luck all around
5961670: This just failed during reimage
5961830: Failed on setup; “SSH connection to gibba016 was lost”
5961831: Failed on setup; “raise AnsibleFailedError(failures)”

Will merge tomorrow once I've figured out how to do proper backports for Pacific and my downstream. 😆
